lgc/patch/PatchInOutImportExport.cpp

```diff
@@ -4568,18 +4484,38 @@ void PatchInOutImportExport::createTessBufferStoreFunction() {
   Value *tfBufferOffset = BinaryOperator::CreateMul(relPatchId, tfByteStride, "", tfStoreBlock);
 
   tfBufferOffset = BinaryOperator::CreateAdd(tfBufferOffset, tfByteOffset, "", tfStoreBlock);
-  tfBufferOffset = BinaryOperator::CreateAdd(tfBufferOffset, tfBufferBase, "", tfStoreBlock);
 
   auto branch = BranchInst::Create(endBlock, tfStoreBlock);
 
+  // Create llvm.amdgcn.raw.tbuffer.store.
+  std::vector<unsigned> formats;
+  if (m_gfxIp.major <= 9) {
+    formats = {
+        ((BUF_NUM_FORMAT_FLOAT << 4) | (BUF_DATA_FORMAT_32)),
+        ((BUF_NUM_FORMAT_FLOAT << 4) | (BUF_DATA_FORMAT_32_32)),
+        ((BUF_NUM_FORMAT_FLOAT << 4) | (BUF_DATA_FORMAT_32_32_32)),
+        ((BUF_NUM_FORMAT_FLOAT << 4) | (BUF_DATA_FORMAT_32_32_32_32)),
+    };
+  } else if (m_gfxIp.major == 10) {
+    formats = {BUF_FORMAT_32_FLOAT, BUF_FORMAT_32_32_FLOAT, BUF_FORMAT_32_32_32_FLOAT, BUF_FORMAT_32_32_32_32_FLOAT};
+  } else
+    llvm_unreachable("Not implemented!");
+
+  std::string callName = "llvm.amdgcn.raw.tbuffer.store." + getTypeName(tfValueTy);
+  CoherentFlag coherent = {};
+  coherent.bits.glc = true;
+
   Value *args[] = {
-      tfValue,                                           // vdata
-      tfBufferDesc,                                      // rsrc
-      tfBufferOffset,                                    // offset
-      ConstantInt::get(Type::getInt32Ty(*m_context), 0), // soffset
-      ConstantInt::get(Type::getInt32Ty(*m_context), 1)  // cachepolicy: glc = 1
+      tfValue,                                  // vdata
+      tfBufferDesc,                             // rsrc
+      tfBufferOffset,                           // voffset
+      tfBufferBase,                             // soffset
+      builder.getInt32(formats[compCount - 1]), // format
+      builder.getInt32(coherent.u32All)         // glc
   };
-  emitCall("llvm.amdgcn.raw.buffer.store.f32", Type::getVoidTy(*m_context), args, {}, branch);
+
+  builder.SetInsertPoint(branch);
+  builder.CreateNamedCall(callName, builder.getVoidTy(), args, {});
```

Comment:
nit: you can use `llvm::Twine` to avoid allocating the string buffer

Reply:
Done.